### ISSUE TITLE
Enhancement around fonts

### ISSUE THREAD

**Issue by mchobby** [ropod7/pyboard_drive#19]
Hi Roman, I did made a new pull request regarding the font usage.
If you are ok with the pull request, I would like to investigate the following points (if you do agree):
- do not load the Arial_14 at LCD startup (letting the developer to chose and load only one font)
- Accept a "font" parameter at LCD() instance creation (would be the default LCD font, None by default to support LCD with only drawing feature)
- reuse the default LCD font with initCh() when "font" parameter is omitted when calling initCh() 
- raise an exception in prinLn() when font is still None (not mentioned at LCD level and not mentioned as initCh parameter)

Regarding the /fonts/xx.py , I would like to investigate for reducing the memory footprint.

**Comment by ropod7** [comment]
I have a trouble with merges. Merge button is useless and github notify me that there is some conflicts.
Otherwise, shore, I am agree.

**Comment by ropod7** [comment]
Good job.

Answers.
1. import line from lcd.py removed.
2. Better if this feature sets up in Widget class. I have to construct it soon. If we use it in LCD class than loss of meaning of initCh method, which give us opportunity to define a lot of string objects with any background colors and font types.
3. same logic.
4. It's done.

My questions.
1. As you see fonts larger than 14 pixels are useless. I pulled up just 14 pixel fonts. Scale by 2 can give us same result.
2. Fonts aligment by bottom pixel must to be fixed. I think we will take as alignment template chars as "g, p, q, Q, y" and put them down if needed, otherwise pick up other chars. Chars as "~ ; ' " we must pick up to. Fonts must to be adjusted.
3. Font bytes from 0 - 31 are deleted, they are useless in PyBoard. They eat a lot of memory.

**Comment by ropod7** [comment]
I will think obout importFont() in BaseChars class. It must to be user friendly.

**Comment by mchobby** [comment]
1. Font bigger make sense when they embed graphics (see the Windings font under windows), heydi_20.py is a graphical font. They are many other free graphical font that does propose such typo (folders, clocks, rings, bells, ...).  To be displayed properly, the font size must be over 14 point (comfortable viewing needs a bigger image rendering).
2. Will have a look to this so the freetype-generator will manage it.
3. I will update freetype-generator to take this option in charge (font like "Windings" may have useful graphical chars)

**Comment by ropod7** [comment]
1. Have you rendered properly heydi_20 in your pyboard using charsTest()? I have not. : )

**Comment by ropod7** [comment]
I will think obout extension for fonts larger than 14. But before, give me to know:
1. Is 20 a largest size, that you will generate?
2. Have you any other sizes between 14 and 20?

**Comment by ropod7** [comment]
It works with height 0x17 (23)!
'height' : 0x17,

It must to be in library!

**Comment by ropod7** [comment]
But, try to set middle point in freetype-generator, if possible. There looks black bold line on the bottom of symbol.
Good. I was thinking obout only 14 px fonts rendering when I was making char printing methods, but it works fine with others to! 
I am surprised!

**Comment by mchobby** [comment]
> I will think obout extension for fonts larger than 14. But before, give me to know:
> 1. Is 20 a largest size, that you will generate?
> 2. Have you any other sizes between 14 and 20?
1. I can generate as big as you which. I always generate numbers as multiple of 8 bits (so 8 points height, 16, 24, 32, etc). 24 pixels may be appropriate for graphical font (32 bits would be a maximum over that this would eat too much memory). I propose to stop on 24 at the moment.
2. I could have more sizes. I will finalize a bit more (and stabilise) the generator code and I will push it to the git. 

> Have you rendered properly heydi_20 in your pyboard using charsTest()?

Not yet but by inspecting the bit this was looking better than rendered on the screen. I will check again.

> But, try to set middle point in freetype-generator, if possible. There looks black bold line on the bottom of symbol.

I will focus on this soon.

**Comment by ropod7** [comment]
Fantastic.
Let's make 24 as largest, than we may change font scale logic. Current give us too thin character, if scale more than 3.

**Comment by mchobby** [comment]
I added the project freetype-generator on GitHub
https://github.com/mchobby/freetype-generator
It contains the generated font is the /ili-fonts/ subfolder.
- There is **new fonts**
- I will now manage the issue about descender chars.

After, I will have a look to the binary storage (into a files).

**Comment by ropod7** [comment]
What about vertical alignment?

**Comment by ropod7** [comment]
Updates:
1. in every font added 'null' key for missing font indexes.
2. added **init**.py with font importing logic.
3. automatic fonts importing by typing args as font='Arial_14'

**Comment by mchobby** [comment]
I just pushed a new version that take care of alignment (descender space). I still have to fix about ~ = ^ and so on... but regular characters are properly aligned.

```
from veram_14 import *

>>> for a_char in ['a','q','y','b']:
...     for value in VeraMono_14[ord(a_char)]:
...         print( bin(value).replace('0',' ').replace('1','*') )
...     print('')
... 
 b*    ***        
 b*   *   * *     
 b*   *   *  *    
 b*   *   *  *    
 b*    *  *  *    
 b*   *******     

 b*     ****      
 b*   **    **    
 b*   *      *    
 b*   *      *    
 b*    *    *     
 b************    

 b*          *    
 b**      ***     
 b***   **        
 b*  ****         
 b*      ***      
 b*         **    

 b*   *********** 
 b*    *    **    
 b*   *      *    
 b*   *      *    
 b*   **    **    
 b*     ****      
```

Now I will manage the remaining of the chars

**Comment by ropod7** [comment]
Cool!

**Comment by mchobby** [comment]
Second push this early morning.
Now it should be completely fixed and all characters (including =,^"+-*~' ... ) properly aligned.

Enjoy

**Comment by mchobby** [comment]
Fixed the font in the ft-generator git.
The font file has been renamed (accordingly to real height of characters).
You can use the following code in the fonts/**init**.py

```
# -*- coding: utf-8 -*-

__all__ = []

def importing(font):
    if font == 'Vera_10':
        from fonts.vera_10 import Vera_10 as font
    elif font == 'Arial_14':
        from fonts.arial_14 import Arial_14 as font
    elif font == 'Vera_15':
        from fonts.vera_15 import Vera_15 as font
    elif font == 'VeraMono_15':
        from fonts.veram_15 import VeraMono_15 as font
    elif font == 'Pitch_15':
        from fonts.pitch_15 import Pitch_15 as font
    elif font == 'Pitch_23':
        from fonts.pitch_23 import Pitch_23 as font
    elif font == 'Vera_23':
        from fonts.vera_23 import Vera_23 as font
    elif font == 'VeraMono_23':
        from fonts.veram_23 import VeraMono_23 as font
    elif font == 'Heydings_23':
        from fonts.heyd_23 import Heydings_23 as font
    elif font == 'Entypo_13':
        from fonts.etypo_13 import Entypo_13 as font
    elif font == 'Entypo_23':
        from fonts.etypo_23 import Entypo_23 as font
    else:
        raise ImportError('wrong font name: ' + font)
    return font
```

Note: I will work on the remaining characters (fix the remaining alignment issue) ... and also remove the unwanted chars.

**Comment by mchobby** [comment]
Fonts have been updated. There is some improvements (you will certainly appreciate them ;-) )
- Font are now pruned (characters without valuable content are removed from the generated python file).
- Font are also produced with minimal character set (just essential characters). They are named: font_base_name + "_m" + font_size_in_pixel + ".py" (note the "m" before the size. "m" for "minimal" charset)

**Comment by mchobby** [comment]
I you know a FreeType font that contains cyrilic characters (eg: on http://www.dafont.com ), I could produce font file for the ILI driver.
Just send the URL for the font :-)

**Comment by mchobby** [comment]
Added the "Arrows" font.
My fonts/**init**.py is currently looking like this

```
# -*- coding: utf-8 -*-

__all__ = []

def importing(font):
    if font == 'Arrows_15':
        from fonts.arrows_15 import Arrows_15 as font
    elif font == 'Arrows_23':
        from fonts.arrows_23 import Arrows_23 as font
    elif font == 'Vera_10':
        from fonts.vera_10 import Vera_10 as font
    elif font == 'Vera_m10': # Minimal Charset
        from fonts.vera_m10 import Vera_m10 as font
    elif font == 'Arial_14':
        from fonts.arial_14 import Arial_14 as font
    elif font == 'Vera_15':
        from fonts.vera_15 import Vera_15 as font
    elif font == 'VeraMono_15':
        from fonts.veram_15 import VeraMono_15 as font
    elif font == 'Pitch_15':
        from fonts.pitch_15 import Pitch_15 as font
    elif font == 'Pitch_23':
        from fonts.pitch_23 import Pitch_23 as font
    elif font == 'Vera_23':
        from fonts.vera_23 import Vera_23 as font
    elif font == 'VeraMono_23':
        from fonts.veram_23 import VeraMono_23 as font
    elif font == 'Heydings_23':
        from fonts.heyd_23 import Heydings_23 as font
    elif font == 'Entypo_13':
        from fonts.etypo_13 import Entypo_13 as font
    elif font == 'Entypo_23':
        from fonts.etypo_23 import Entypo_23 as font
    else:
        raise ImportError('wrong font name: ' + font)
    return font
```

**Comment by mchobby** [comment]
Rotating arrow (based on Arrows_23 font).
Not perfect because background is not cleaned up before printing... but give a nice effect.

```
import pyb
from lcd import *
l = LCD( rate=21000000)
l.fillMonocolor( GREEN )
c = l.initCh(font='Arrows_23', color=BLACK, bgcolor=GREEN)

wait_seq = [99,103,97,104,100,102,98,101]

for i in range( 10 ): # Repeat 10 times
   for ordinal in wait_seq:
      c.printChar( chr(ordinal),10,10 )
      pyb.delay( 100 )
```

Could also be tried with the following sequence (gives a better rotating result).

wait_seq = [67,71,65,72,68,70,66,69]

**Comment by ropod7** [comment]
Look, there in micropython forum Damien give us a nice solution, we will look forward to play with it.
http://forum.micropython.org/viewtopic.php?f=6&t=1776
But unfortunatelly, it's not so user friendly, if we compile our own firmware with our modules. Somewhere it should be helpful.
For example we may compile all in the ILI9341 and try to check performance and memory allocation.
This evening I will play with it.

PS.
Fonts are famous. Thank you.

**Comment by mchobby** [comment]
Hi Roman,

I did indeed noticed this feature,
It was adviced when I did open a forum thread for "font storage and 
minimized memory consumption".

However, this approach would made the driver code useless if you did not 
have a proper firmware burned on the PyBoard.
This solution would be welcome for a dedicated hardware plateform 
(screen+pyboard+...)... but not suitable for a "ready to use" driver code.

If memory would be a problem then I would be in favor of  reading the 
font from a binary file... because this approach still use the standard 
micropython firmware (the driver would be more portable).

This week-end, I will have a look to the binary storage (the bytearray) 
for the font.

Kind regards,
Dominique

Meurisse D. (Gérant)
MC Hobby SPRL
Service R&D et support
